lib/AST/AccessNotes.cpp

```diff
@@ -0,0 +1,295 @@
+  dump(llvm::dbgs());
+  llvm::dbgs() << "\n";
+}
+void AccessNote::dump() const {
+  dump(llvm::dbgs());
+  llvm::dbgs() << "\n";
+}
+
+void AccessNotes::dump(llvm::raw_ostream &os) const {
+  os << "(access_notes reason='" << reason << "'";
+  for (const auto &note : notes) {
+    os << "\n";
+    note.dump(os, /*indent=*/2);
+  }
+}
+
+void AccessNote::dump(llvm::raw_ostream &os, int indent) const {
+  os.indent(indent) << "(note name='";
+  name.print(os);
+  os << "'";
+
+  if (name.name.getBaseName().isSpecial())
+    os << " is_special_name";
+
+  if (ObjC)
+    os << " objc=" << *ObjC;
+  if (ObjCName)
+    os << " objc_name='" << *ObjCName << "'";
+  if (Dynamic)
+    os << " dynamic=" << *Dynamic;
+
+  os << ")";
+}
+
+}
+
+LLVM_YAML_DECLARE_SCALAR_TRAITS(swift::AccessNoteDeclName, QuotingType::Single);
+LLVM_YAML_DECLARE_SCALAR_TRAITS(swift::ObjCSelector, QuotingType::Single);
+LLVM_YAML_IS_SEQUENCE_VECTOR(swift::AccessNote)
+LLVM_YAML_DECLARE_MAPPING_TRAITS(swift::AccessNotes)
+
+// Not using macro to avoid validation issues.
+template <> struct llvm::yaml::MappingTraits<swift::AccessNote> {
+  static void mapping(IO &IO, swift::AccessNote &Obj);
+  static StringRef validate(IO &IO, swift::AccessNote &Obj);
+};
+
+namespace swift {
+llvm::Expected<AccessNotes>
+AccessNotes::load(ASTContext &ctx, llvm::MemoryBuffer *buffer) {
+  llvm::yaml::Input yamlIn(buffer->getBuffer(), (void *)&ctx);
+
+  AccessNotes notes;
+  yamlIn >> notes;
+
+  if (yamlIn.error())
+    return llvm::errorCodeToError(yamlIn.error());
+
+  return notes;
+}
+}
+
+
+namespace llvm {
+namespace yaml {
+
+using AccessNote = swift::AccessNote;
+using AccessNotes = swift::AccessNotes;
+using ASTContext = swift::ASTContext;
+using AccessNoteDeclName = swift::AccessNoteDeclName;
+using ObjCSelector = swift::ObjCSelector;
+
+void ScalarTraits<AccessNoteDeclName>::
+output(const AccessNoteDeclName &name, void *ctxPtr, raw_ostream &os) {
+  name.print(os);
+}
+
+StringRef ScalarTraits<AccessNoteDeclName>::
+input(StringRef str, void *ctxPtr, AccessNoteDeclName &name) {
+  ASTContext &ctx = *static_cast<ASTContext *>(ctxPtr);
+  name = AccessNoteDeclName(ctx, str);
+  return name.empty() ? "invalid declaration name" : "";
+}
+
+void ScalarTraits<ObjCSelector>::output(const ObjCSelector &selector,
+                                        void *ctxPtr, raw_ostream &os) {
+  os << selector;
+}
+
+StringRef ScalarTraits<ObjCSelector>::input(StringRef str, void *ctxPtr,
+                                            ObjCSelector &selector) {
+  ASTContext &ctx = *static_cast<ASTContext *>(ctxPtr);
+
+  if (auto sel = parseObjCSelector(ctx, str)) {
+    selector = *sel;
+    return "";
+  }
+
+  return "invalid selector";
+}
+
+void MappingTraits<AccessNote>::mapping(IO &io, AccessNote &note) {
+  io.mapRequired("Name", note.name);
+  io.mapOptional("ObjC", note.ObjC);
+  io.mapOptional("Dynamic", note.Dynamic);
+  io.mapOptional("ObjCName", note.ObjCName);
```

Comment:
Can we make this ignore unknown keys instead of erroring out?

Reply:
Without modifying `llvm::yaml::Input`, the best we can do is mark unused keys as used (to suppress the error) and add them to a list to be diagnosed.